### PR TITLE
Windows: Retry workaround for RS1/RS2 compute system enumeration

### LINKED ISSUE #30278: docker rm throws access denied error after upgrade to new version of docker
this ticket is a copy of [#374 in docker/for-win](https://github.com/docker/for-win/issues/374)
more details and additional comments can be viewed over there.
@simonferquel, @friism see this ticket with docker-engine.

### Expected behavior
a `docker rm {CID}` does not throw a access denied error while renaming the container to "{CID}_removing".

### Actual behavior
removing a container with `docker rm` outputs:

```
λ docker ps -a                                                                                                                                                                                                  
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES                                                                                   
69b66d310c2e        python-build        "powershell"        5 weeks ago         Dead                                    furious_nightingale                                                                     
389bdb07a55b        python-build        "powershell"        5 weeks ago         Dead                                    focused_volhard                                                                         
                                                                                                                                                                                                                
D:\Entwicklung\docker-w-python                                                                                                                                                                                  
λ docker rm -f 69                                                                                                                                                                                               
Error response from daemon: Driver windowsfilter failed to remove root filesystem 69b66d310c2e7c6202a633dd5d88ea5cb9ee390b2d98ff299d4b1fb0258ff02e: rename E:\Docker\windowsfilter\69b66d310c2e7c6202a633dd5d88e
a5cb9ee390b2d98ff299d4b1fb0258ff02e E:\Docker\windowsfilter\69b66d310c2e7c6202a633dd5d88ea5cb9ee390b2d98ff299d4b1fb0258ff02e-removing: Zugriff verweigert       
```                                                
                                                                                                                                                                                                      

### Information

**λ docker info**                                                                                 
```
Containers: 1                                                                                 
 Running: 0                                                                                   
 Paused: 0                                                                                    
 Stopped: 1                                                                                   
Images: 29                                                                                    
Server Version: 1.13.0-rc5                                                                    
Storage Driver: windowsfilter                                                                 
 Windows:                                                                                     
Logging Driver: json-file                                                                     
Plugins:                                                                                      
 Volume: local                                                                                
 Network: l2bridge l2tunnel nat null overlay transparent                                      
Swarm: inactive                                                                               
Default Isolation: hyperv                                                                     
Kernel Version: 10.0 14393 (14393.576.amd64fre.rs1_release_inmarket.161208-2252)              
Operating System: Windows 10 Pro                                                              
OSType: windows                                                                               
Architecture: x86_64                                                                          
CPUs: 8                                                                                       
Total Memory: 11.99 GiB                                                                       
Name: BLAUBEERE10                                                                             
ID: E2WR:C7YS:2BZ5:TNFX:NCJL:FZPY:6U6F:3MFC:CBX4:DJCP:LZFO:RZS3                               
Docker Root Dir: E:\Docker                                                                    
Debug Mode (client): false                                                                    
Debug Mode (server): true                                                                     
 File Descriptors: -1                                                                         
 Goroutines: 18                                                                               
 System Time: 2017-01-08T15:31:06.1621483+01:00                                               
 EventsListeners: 0                                                                           
Http Proxy: 
Https Proxy:  
No Proxy:                   
Registry: https://index.docker.io/v1/                                                         
Experimental: true                                                                            
Insecure Registries:                                                                          
 127.0.0.0/8                                                                                  
Live Restore Enabled: false                                                                   
```

**λ docker version**
```
Client:
 Version:      1.13.0-rc5
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 03:07:30 2017
 OS/Arch:      windows/amd64

Server:
 Version:      1.13.0-rc5
 API version:  1.25 (minimum version 1.24)
 Go version:   go1.7.3
 Git commit:   43cc971
 Built:        Thu Jan  5 03:07:30 2017
 OS/Arch:      windows/amd64
 Experimental: true
```



### Steps to reproduce the behavior

  1. docker image created with previous docker version
  2. docker rm {CID}

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>

Fixes https://github.com/docker/docker/issues/30278

@georgyturevich @cons0l3 FYI

There's a race condition in HCS in RS1 and RS2 builds (fixed in RS3) whereby the call to
enumerate compute systems can return access denied if a silo is being torn down in the kernel
underneath it.

The net result is that `docker rm` may fail. In fact, using these two scripts, I have been
able to repro the failure almost immediately (usually with less than 10 containers having 
been started)

```
docker stop recovertest
docker rm -f recovertest
remove-item "failed.txt" -erroraction silentlycontinue
while ($true) {
    $count++
    Write-Host "Iteration $count $(Get-Date)"
    docker ps -q
    docker run -d -i --name recovertest microsoft/iis
    if ($lastexitcode -ne 0) { new-item ".\failed.txt"; exit -1 }
    docker stop recovertest
    if ($lastexitcode -ne 0) { new-item ".\failed.txt"; exit -1 }
    docker rm recovertest
   if ($lastexitcode -ne 0) { new-item ".\failed.txt"; exit -1 }
   if (Test-Path ".\failed.txt") { Write-Host "Quitting as failed.txt exists"; exit -1 }
}
```

```
docker stop recovertest2
docker rm -f recovertest2
remove-item "failed.txt" -erroraction silentlycontinue
while ($true) {
    $count++
    Write-Host "Iteration $count $(Get-Date)"
    docker ps -q
    docker run -d -i --name recovertest2 microsoft/iis
    if ($lastexitcode -ne 0) { new-item ".\failed.txt"; exit -1 }
    docker stop recovertest2
    if ($lastexitcode -ne 0) { new-item ".\failed.txt"; exit -1 }
    Start-Sleep -Seconds 5
    docker rm recovertest2
   if ($lastexitcode -ne 0) { new-item ".\failed.txt"; exit -1 }
   if (Test-Path ".\failed.txt") { Write-Host "Quitting as failed.txt exists"; exit -1 }
}
```

With this fix, I've had the scripts running for a couple of hours without failure.

@darrenstahlmsft @johnstep PTAL.

@thaJeztah @vieux @friism - this is one that should be strongly considered for a future 1.13 release, and for CS.

### CODE REVIEW

File: daemon/graphdriver/windows/windows.go [hunk 1]
```diff
@@ -265,19 +266,35 @@ func (d *Driver) Remove(id string) error {
 	// it is a transient error. Retry until it succeeds.
 	var computeSystems []hcsshim.ContainerProperties
 	retryCount := 0
+	osv := system.GetOSVersion()
 	for {
-		// Get and terminate any template VMs that are currently using the layer
+		// Get and terminate any template VMs that are currently using the layer.
+		// Note is unfortunate that we end up in the graphdrivers Remove() call
```
Comment (johnstep) [inline daemon/graphdriver/windows/windows.go]: Nit: Replace `Note is` with something like `Note: It is`.

**Review by johnstep — APPROVED** [review]
LGTM

File: daemon/graphdriver/windows/windows.go (hunk 1 above)
Comment (jhowardmsft) [inline daemon/graphdriver/windows/windows.go]: Updated

**Comment by darrenstahlmsft** [comment]
Not a maintainer, but LGTM

**Comment by vieux** [comment]
LGTM

**Pull request merged**